### DISCUSSION TITLE
Table in PDF on custom page not working

### DISCUSSION THREAD

**Discussion by griggsk** [filamentphp/filament#12562] [Help] [answered]
### Package

Table builder

### Package Version

v3.0.0

### How can we help you?

I have a custom page with a table that I am trying to download as a PDF. The custom page includes a filament table and some other content. I can generate the PDF without the table and I can view the page with the table. But I cannot generate a PDF with the table. 

I am asking how I can generate a PDF with a filament table. 
1. I am using Barryvdh\DomPDF\Facade\Pdf
2. I know that the styling might not all work. I just want the HTML for the table so I can reuse the code from my view. 
3. Issue: I get the error " Using $this when not in object context" on the $this->table. I understand why I get this error, but how can I get the table. 

Q1: Is it possible to use a filament table in a PDF? 
Q2: Is it possible to get the HTML from a table?
Q3: Any other ideas on how to do this?

Thanks. 

See Code: 
```
<?php

namespace App\Livewire\Response;

use App\Models\Response;
use Filament\Forms\Concerns\InteractsWithForms;
use Filament\Forms\Contracts\HasForms;
use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Concerns\InteractsWithTable;
use Filament\Tables\Contracts\HasTable;
use Filament\Tables\Table;
use Illuminate\Contracts\View\View;
use Illuminate\Support\Facades;
use Livewire\Component;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Barryvdh\DomPDF\Facade\Pdf;

class DownloadResponse extends Component implements HasForms, HasTable
{

    use InteractsWithTable;
    use InteractsWithForms;

   
    public Response $response;
  
    
    public function table(Table $table): Table
    {
      return $table
            ->relationship(fn (): HasMany => $this->response->summaries())
            ->inverseRelationship('response')
            ->columns([
               TextColumn::make('domain.name'), 
               TextColumn::make('status.label')->badge(),
            ])
            ->paginated(false);
    }
    
   

    public function download()
    {

         $pdf = Pdf::loadView('livewire.response.download', 
            [
              'response' => $this->response,
              'summaries' => $this->response->summaries,
            ])->output();
        
        

        return response()->streamDownload(
             function ()  use($pdf) {echo $pdf;},
             "filename.pdf"
        );

    }

    public function render()
    {
        return view('livewire.response.download', 
            [
              'response' => $this->response,
              'summaries' => $this->response->summaries,
            ])->layout('layouts.guest');

    }
}

```
```
 Route

Route::get('download/download/{response}', DownloadResponse::class, 'download')->name('download');

download.blade.php 
<html>
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
    <title>OSP</title>
</head>
<body>
	<div>
		<p>Child Initial: {{$response->child->initials }}</p>
		{{ $this->table }}
		
	</div>
</body>
</html>		
```

**Comment by griggsk** [comment]
A few new things I have tried

`$table = $this->getTableRecords();` and passing this to the view
This works but doesn't get me anything useful

I thought this was promising - but it's returning the same error on the to HTML method: [Using $this when not in object context]
```
$table = $this->getTable();
 dd($table->toHtml());
```

Any ideas?

**Comment by griggsk** [comment]
Can someone explain why $this->table->getHTML() doesn't work? I can do $this->table->render() and see that $this is the table.

**Comment by xalabama** [comment]
The reason is that the view file that is rendered belongs to an livewire component and uses $this inside. You can have a look to this if you vendor publish the filament-tables views and have a look into the index.blade.php. Put simply you can render the livewire component inside a blade file but not outside of its context. 

But there is a workaround for this. You could render your created livewire component DownloadResponse inside a blade view file and then render the blade view inside of your method. 

For example:

Your livewire view file:
```
<div>
    {{ $this->table }}
</div>
```

Your normal blade view file, lets call it plain-table.blade.php
```
<livewire:your-component-name />
```

Your method call:
```
public function getTableAsHtml(): string
{
    return view('plain-table')->render(); 
}
```

The result will be the html string from your plain-table.blade.php within your table. Because you want to render this inside a PDF file, this solution will also helps you to style your table how you want. 

I hope this will help you!

**Reply by griggsk** [reply]
Thank you for that explanation. I've made a lot of progress thanks to you! 
I can get the table but it has no data. It's not getting the Response Model that is mounted in the view.
I tried binding the response model in the plain table view and the method but that gives an underfined error. 

Any more ides? Thanks
```
class DownloadResponse extends Component implements HasForms, HasTable
{

    use InteractsWithTable;
    use InteractsWithForms;

   
    public Response $response;
    public function mount(Response $response)
    {
        $this->response = $response;
    }
    
    public function table(Table $table): Table
    {
      return $table
            ->relationship(fn (): HasMany => $this->response->summaries())
            ->inverseRelationship('response')
            ->columns([
               TextColumn::make('domain.name'), 
               TextColumn::make('status.label')->badge(),
            ])
            ->paginated(false);
    }
    
    public function download()
    {
         $records = $this->getTableAsHtml();
         dd($records);
    }

    public function getTableAsHtml(): string
    {
        return view('livewire.response.plain-table')->render(); 
    }

    public function render()
    {
        return view('livewire.response.download', 
            [
              'response' => $this->response,
            ])->layout('layouts.guest');

    }
}

plain-table.blade.pho
<livewire:response.download-response/>

download.blade.php
<div>
{{ $this->table }}
</div>
```

**Reply by xalabama** [reply, marked as answer]
If i understand your code correct you want to call the response model from a parameter of your route? Im not sure how this will possible but i prefer to provide something like an identifier for the model that has to initialize. Im also not knowing your model structure, so i only could provide you how i can solve this successfully with a user model in relation to his roles. 

Than it looks like this:
```
// ...
   
    // Param that is holding the model
    public User $user;
    
    public function mount(int $userId)
    {
        // Initiate instance of the model
        $this->user = User::query()->find($userId);
    }

    public function table(Table $table): Table
    {
        return $table
            ->relationship(fn(): BelongsToMany => $this->user->roles())
            ->inverseRelationship('users')
            ->columns([
                TextColumn::make('name')->badge(), // the name of the role 
            ])
            ->paginated(false);
    }

// ... 
```

But than you get a problem with your concept because you cant call the download method before initializing the the table method that needs the initialized model from the mount lifecycle method. Hmm. 

Finally i would suggest to call the getTableAsHtml function content inside a different controller. I keep my example with the user model.

Route
```
Route::get('download/download/{userId}', [\App\Http\Controllers\DownloadController::class, 'download'])->name('download');
```

DownloadController
```
class DownloadController extends Controller
{
    public function download(Request $request, $userId): string
    {
        $data = view('livewire.plain-table', ['userId' => $userId])->render();

        dd($data);
    }
}
```

Table
```
// ...

public User $user;

    public function mount(int $userId)
    {
        $this->user = User::query()->find($userId);
    }

    public function table(Table $table): Table
    {
        return $table
            ->relationship(fn(): BelongsToMany => $this->user->roles())
            ->inverseRelationship('users')
            ->columns([
                TextColumn::make('name')->badge(),
            ])
            ->paginated(false);
    }

    public function render()
    {
        return view('livewire.response.download', [
        // ... 
        ]);
    }

// ...
```

plain-table.blade.php
```
<livewire:response.download-response :userId="$userId" />
```

Than you can keep your download logic inside the DownloadController and the table logic inside the livewire component. Don't forget to protect the DownloadController route with middleware if it should not be public.

I hope have understood you correct and this is what you are looking for. If not feel free to explain your concept. ;)

**Reply by griggsk** [reply]
Yes! Thank you so much. Moving the download into it's own controller was the key.  

Here is the full code in case it's helpful to others.

**Routes**
`Route::get('download/view/{response}', DownloadResponse::class);`
`Route::get('download/response/{response}', [DownloadController::class, 'download'])->name('download');
`

**DownloadResponse**
```
class DownloadResponse extends Component implements HasForms, HasTable
{

    use InteractsWithTable;
    use InteractsWithForms;

    public Response $response;
    public function mount(Response $response)
    {
        $this->response = $response;
    }
    
    public function table(Table $table): Table
    {
      return $table
            ->relationship(fn (): HasMany => $this->response->summaries())
            ->inverseRelationship('response')
            ->columns([
               TextColumn::make('domain.name'), 
               TextColumn::make('status.label')->badge(),
            ])
            ->paginated(false);
    }
    
    public function render()
    {
        return view('livewire.response.download', 
            [
              'response' => $this->response,
            ])->layout('layouts.guest');

    }
```
**download.blade**
```
	<div>
		<p>Child Initial: {{$response->child->initials }}</p>
		{{ $this->table}}
	</div>
```
**DownloadController**
```
class DownloadController extends Controller
{
    public function download(Request $request, Response $response)
    {
        $pdf = Pdf::loadView('livewire.response.plain-table', 
            [
                'response' => $response
            ]
            )->output();
        

        return response()->streamDownload(
             function ()  use($pdf) {echo $pdf;},
             "filename.pdf",
            ['Content-Type' => 'application/pdf']
        );
    }
```
**plain-table.blade**
```
<livewire:response.download-response :response="$response"/>
```
